libcst/_nodes/internal.py

```diff
@@ -70,6 +71,30 @@ def record_syntactic_position(
         yield
 
 
+@add_slots
+@dataclass(frozen=False)
+class CodegenWriter(CodegenState):
```

Comment:
Seems like the class hierarchy should be the other way around. `CodegenState` should specialize `CodegenWriter`.

Then we wouldn't have to have an awkward default value for `writer`.